aten/src/ATen/CUDAStreamGuard.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <ATen/CUDAStream.h>
+#include <ATen/Context.h>
+
+#include <cstddef>
+
+namespace at {
+
+/// RAII guard that sets the current CUDA Stream for a particular CUDA device,
+/// identified by its device index, and sets it back to the original stream upon
+/// destruction.
+///
+/// The stream is always reset to the one that was active at the time of
+/// construction of the guard. Even if you `set_stream` after construction, the
+/// destructor will still reset the stream to the one that was active at
+/// construction time.
+struct CUDAStreamGuard {
+  /// Default constructor, does nothing and causes no change in the current
+  /// stream until `set_stream` is called.
+  CUDAStreamGuard() = default;
+
+  /// Sets the CUDA stream on the given CUDA device (calls `set_stream`).
+  CUDAStreamGuard(int32_t device_index, const CUDAStream& stream) {
+    set_stream(device_index, stream);
+  }
+
+  /// Resets the CUDA stream on the stored device to the one that was active at
+  /// the time of construction of this object.
+  ~CUDAStreamGuard() {
+    if (original_stream_) {
+      globalContext().uncheckedSetCurrentCUDAStreamOnDevice(
+          device_index_, original_stream_);
+    }
+  }
+
+  /// Sets the CUDA stream on the CUDA device identified by the given
+  /// `device_index` to the one specified by `stream`.
+  void set_stream(int32_t device_index, const CUDAStream& stream) {
```

Comment:
Two questions:

- Should the device_index be int64_t instead of int32_t? I think @ezyang suggested the former was appropriate for cases like this.
- Instead of creating separate CUDAStreamGuard and CUDADeviceGuard and probably CUDADeviceAndStreamGuard classes, does it make sense to create a single CUDAGuard class that can accept setting the device or stream or both? I actually need to set both in #8354, and at the time created an AutoGPUStream (based on the existing AutoGPU and AutoStream). I'd like to get rid of that (obviously) and use a class like this, and I think the logic is so similar between these three use cases that they are likely best combined into one class.

Reply:
Looking at the removed auto_stream.h later, I like creating a single CUDAGuard class even more. auto_stream.h was was not safe to use if when changing the GPU during the guard's lifetime, and you've done a good job respecting the stream<->device relationship with this new class. It seems natural (to me) to also add a setDevice() to it.

Reply:
As for `int32_t` vs. `int64_t` -- I went for `int32_t` in `at::Device` because in CUDA-land it's just `int`, which is usually 32 bits (but I prefer fixed width types). That said, extending it to 64 bit should be pretty harmless right now if we wanted to changed it. Any opinions @ezyang @mruberry?